server/src/main/java/org/opensearch/search/SearchService.java

```diff
@@ -788,28 +856,57 @@ final ReaderContext createAndPutReaderContext(
      * Opens the reader context for given shardId. The newly opened reader context will be keep
      * until the {@code keepAlive} elapsed unless it is manually released.
      */
-    public void openReaderContext(ShardId shardId, TimeValue keepAlive, ActionListener<ShardSearchContextId> listener) {
-        checkKeepAliveLimit(keepAlive.millis());
+    public void createPitReaderContext(ShardId shardId, TimeValue keepAlive, ActionListener<ShardSearchContextId> listener) {
```

Comment:
Is it possible to re-work this to leverage Lucene's [`SearcherLifetimeManager`](https://github.com/apache/lucene/blob/main/lucene/core/src/java/org/apache/lucene/search/SearcherLifetimeManager.java) to handle all of the lifecycle logic?

Reply:
@nknize Thanks for the great suggestion. Since most of the changes to support PIT is almost ready, do you think we can get to refactoring as a fast follow up once we have segrep in place. The current changes aim to simplify the problems pertaining to scrolls(retry failures/resiliency) and supports reuse of the same context across multiple reads which has some really nice improvements(around resource optimization) over the traditional scroll.

Reply:
@nknize We will take up the refactoring once the segrep changes are in place. For now, we will be going ahead with the current implementation in this PR. Let us know if you have any further comments on this.

Reply:
> Since most of the changes to support PIT is almost ready, do you think we can get to refactoring as a fast follow up once we have segrep in place.

I'm +1 to this. I'm not intending for you to overhaul this PR for now - it's pretty large as it is and I like the progress toward unleashing PIT w/o holding it up.